src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -0,0 +1,250 @@
+      if (thread != threads.end()) {
+        if (thread->second.retry_oom_injected) {
+          thread->second.retry_oom_injected = false;
+          JNIEnv *env = cudf::jni::get_jni_env(jvm);
+          // TODO cache what is needed for this...
+          jclass ex_class = env->FindClass("com/nvidia/spark/rapids/jni/RetryOOM");
+          if (ex_class != nullptr) {
+            env->ThrowNew(ex_class, "OOM injected");
+          }
+          throw cudf::jni::jni_exception("injected RetryOOM");
+        }
+
+        if (thread->second.split_and_retry_oom_injected) {
+          thread->second.split_and_retry_oom_injected = false;
+          JNIEnv *env = cudf::jni::get_jni_env(jvm);
+          // TODO cache what is needed for this...
+          jclass ex_class = env->FindClass("com/nvidia/spark/rapids/jni/SplitAndRetryOOM");
+          if (ex_class != nullptr) {
+            env->ThrowNew(ex_class, "OOM injected");
+          }
+          throw cudf::jni::jni_exception("injected SplitAndRetryOOM");
+        }
+      }
+    }
+    return resource->allocate(num_bytes, stream);
+  }
+
+  void do_deallocate(void *p, std::size_t size, rmm::cuda_stream_view stream) override {
+    resource->deallocate(p, size, stream);
+  }
+
+  std::pair<size_t, size_t> do_get_mem_info(rmm::cuda_stream_view stream) const override {
+    return resource->get_mem_info(stream);
+  }
+};
+
+} // empty namespace
+
+extern "C" {
+
+JNIEXPORT jlong JNICALL Java_com_nvidia_spark_rapids_jni_SparkResourceAdaptor_getCurrentThreadId(
+        JNIEnv *env, jclass) {
+  try {
+    cudf::jni::auto_set_device(env);
+    return static_cast<jlong>(pthread_self());
+  }
+  CATCH_STD(env, 0)
+}
+
+JNIEXPORT jlong JNICALL Java_com_nvidia_spark_rapids_jni_SparkResourceAdaptor_createNewAdaptor(
+        JNIEnv *env, jclass, jlong child) {
+  JNI_NULL_CHECK(env, child, "child is null", 0);
+  try {
+    cudf::jni::auto_set_device(env);
+    auto wrapped = reinterpret_cast<rmm::mr::device_memory_resource *>(child);
+    auto ret = new spark_resource_adaptor(env, wrapped);
+    return cudf::jni::ptr_as_jlong(ret);
+  }
+  CATCH_STD(env, 0)
+}
+
+JNIEXPORT void JNICALL Java_com_nvidia_spark_rapids_jni_SparkResourceAdaptor_releaseAdaptor(
+        JNIEnv *env, jclass, jlong ptr) {
+  try {
+    cudf::jni::auto_set_device(env);
+    auto mr = reinterpret_cast<spark_resource_adaptor *>(ptr);
+    delete mr;
+  }
+  CATCH_STD(env, )
+}
+
+JNIEXPORT void JNICALL Java_com_nvidia_spark_rapids_jni_SparkResourceAdaptor_associateThreadWithTask(
+        JNIEnv *env, jclass, jlong ptr, jlong thread_id, jlong task_id) {
+  try {
+    cudf::jni::auto_set_device(env);
+    auto mr = reinterpret_cast<spark_resource_adaptor *>(ptr);
+    mr->associate_thread_with_task(thread_id, task_id);
+  }
+  CATCH_STD(env, )
+}
+
+JNIEXPORT void JNICALL Java_com_nvidia_spark_rapids_jni_SparkResourceAdaptor_removeThreadAssociation(
+        JNIEnv *env, jclass, jlong ptr, jlong thread_id) {
```

Comment:
For `ptr` cast here and in the next two functions, should we use `JNI_NULL_CHECK`?